### PR TITLE
warning mailer for mentors/mentees missing from website

### PR DESCRIPTION
**Pull request by anjali-sharma**

### CODE REVIEW

File: app/models/user.rb [hunk 1]
```diff
@@ -46,6 +46,16 @@ def last_week_tasks(date)
     project.tasks.where(updated_at: (date.beginning_of_week..date.end_of_week))
   end
 
+  def self.mentees_missing_on_website
+    mentees_missing = User.mentee.where("last_activity_at is null or last_activity_at >= (?)", 15.days.ago)
```
Comment (DMips) [inline app/models/user.rb]: @anjali-sharma just noticed - it's a scope so you don't need to use `User` just `mentee.where...` is enough.

Also let's not store results in variables but just connect where and where.not like `mentee.where(...).where.not(..)`
same in mentors

File: app/models/user.rb [hunk 2]
```diff
@@ -46,6 +46,16 @@ def last_week_tasks(date)
...
+    mentees_missing.where.not("is_missing = (?) OR (send_warning_email_after is not null AND send_warning_email_after > (?))", true, Date.today)
+  end
+
+  def self.mentors_missing_on_website
+    mentors_missing = User.mentor.where("last_activity_at is null or last_activity_at >= (?)", 15.days.ago)
+    mentors_missing.where.not("is_missing = (?) OR (send_warning_email_after is not null AND send_warning_email_after > (?))", true, Date.today)
```
Comment (DMips) [inline app/models/user.rb]: @anjali-sharma We forgot about one more thing - we also shouldn't send warning emails to people that were just matched. So we need to first check when the pair was created (sometimes we match existing users so checking just `last_activity_at` won't be enough).

Comment (anjali-sharma) [inline app/models/user.rb]: @DMips alright... shall look into that today. :+1:

Comment (anjali-sharma) [inline app/models/user.rb]: @DMips how much time should I compare created_at for ApplicationMatch to, for being skipping warning emails?

Comment (DMips) [inline app/models/user.rb]: @anjali-sharma hmm I think 10 days will be enough for them to get in touch and start working on the project

Comment (anjali-sharma) [inline app/models/user.rb]: @DMips ok thanks. :+1:

**Comment by anjali-sharma** [comment]
@DMips Have made the needed updates.

**Comment by DMips** [comment]
@anjali-sharma I'll take a look at it later today :)

File: app/models/project.rb [hunk 3]
```diff
@@ -36,4 +36,16 @@ def self.get_users_missing_project_setup
...
+
+  def pair_matched_at
+    pair_matched_at = nil
+    if self.mentee.present?
+      mentee_application = MenteeApplication.find_by_email(self.mentee.email)
+      if mentee_application.present?
+        pair_matched_at = ApplicationMatch.find_by(mentee_application_id: mentee_application.id).created_at
+      end
+    end
+    binding.pry
```
Comment (DMips) [inline app/models/project.rb]: @anjali-sharma you forgot to remove binding ;)

File: app/models/user.rb [hunk 4]
```diff
@@ -53,6 +53,37 @@ def midterm_self_evaluation
     mentee? ? mentee_midterm_evaluation : mentor_midterm_evaluation
   end
 
+  def self.mentees_missing_on_website
+    mentees_missing = User.mentee.where("last_activity_at is null or last_activity_at <= (?)", 15.days.ago)
+    mentees_missing = mentees_missing.where.not("is_missing = (?) OR (send_warning_email_after is not null AND send_warning_email_after > (?))", true, Date.today)
+    # Exclude mentees just matched or rematched
+    projects = Project.where("mentee_id IN (?)", mentees_missing.pluck(:id))
+    mentees_just_matched = []
+    binding.pry
```
Comment (DMips) [inline app/models/user.rb]: @anjali-sharma same here ;)

File: app/models/user.rb [hunk 5]
```diff
@@ -53,6 +53,37 @@ def midterm_self_evaluation
...
+    projects.each do |project|
+      if project.pair_matched_at >= 10.days.ago
+        mentees_just_matched << project.mentee_id
+      end
+    end
+    binding.pry
```
Comment (DMips) [inline app/models/user.rb]: @anjali-sharma

File: app/models/user.rb [hunk 6]
```diff
@@ -53,6 +53,37 @@ def midterm_self_evaluation
...
+    mentors_missing = mentors_missing.where.not("is_missing = (?) OR (send_warning_email_after is not null AND send_warning_email_after > (?))", true, Date.today)
+    # Exclude mentors just matched or rematched
+    projects = Project.where("mentor_id IN (?)", mentors_missing.pluck(:id))
+    mentors_just_matched = []
+    projects.each do |project|
+      if project.pair_matched_at >= 10.days.ago
+        mentors_just_matched << project.mentor_id
+      end
+    end
+    binding.pry
```
Comment (DMips) [inline app/models/user.rb]: @anjali-sharma binding

Comment (anjali-sharma) [inline app/models/user.rb]: Ahh.... too much debugging I did.. :P 
will just remove it !

**Pull request merged**